src/NServiceBus.Core/Pipeline/Outgoing/OutgoingContext.cs

```diff
@@ -11,10 +11,11 @@ protected OutgoingContext(string messageId, Dictionary<string, string> headers,
             : base(parentContext)
         {
             Headers = headers;
-            MessageId = messageId;
+
+            Headers[NServiceBus.Headers.MessageId] = messageId;
```

Comment:
Good catch. So basically then we need to check if it is set and only set it when there is no such header. That should work